mm2src/coins/lightning/ln_utils.rs

```diff
@@ -1,115 +1,97 @@
-use rand::RngCore;
-use std::convert::TryInto;
-use std::net::{IpAddr, Ipv4Addr};
-use std::path::PathBuf;
-use std::sync::Arc;
-use std::time::SystemTime;
-use tokio::net::TcpListener;
-
-const CHECK_FOR_NEW_BEST_BLOCK_INTERVAL: u64 = 60;
-const BROADCAST_NODE_ANNOUNCEMENT_INTERVAL: u64 = 60;
+use derive_more::Display;
+use secp256k1::PublicKey;
+use std::collections::HashMap;
+use std::fs::{File, OpenOptions};
+use std::io::{BufRead, BufReader, Read, Write};
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
 
+cfg_native! {
+    use crate::utxo::rpc_clients::{electrum_script_hash, BestBlock as RpcBestBlock, ElectrumBlockHeader, ElectrumClient,
+                                   ElectrumNonce, UtxoRpcError};
+    use crate::utxo::sign_tx;
+    use bitcoin::blockdata::block::BlockHeader;
+    use bitcoin::blockdata::constants::genesis_block;
+    use bitcoin::blockdata::script::Script;
+    use bitcoin::blockdata::transaction::{Transaction, TxOut};
+    use bitcoin::consensus::encode::deserialize;
+    use bitcoin::hash_types::{BlockHash, TxMerkleNode, Txid};
+    use bitcoin_hashes::{sha256d, Hash};
+    use common::executor::{spawn, Timer};
+    use common::ip_addr::fetch_external_ip;
+    use common::jsonrpc_client::JsonRpcErrorType;
+    use common::{block_on, log};
+    use common::log::LogState;
+    use futures::compat::Future01CompatExt;
+    use futures::lock::Mutex as AsyncMutex;
+    use lightning::chain::keysinterface::{InMemorySigner, KeysInterface, KeysManager};
+    use lightning::chain::transaction::OutPoint;
+    use lightning::chain::{chainmonitor, Access, BestBlock, Confirm, Filter, Watch, WatchedOutput};
+    use lightning::ln::channelmanager;
+    use lightning::ln::channelmanager::{ChainParameters, ChannelManagerReadArgs, SimpleArcChannelManager};
+    use lightning::ln::msgs::NetAddress;
+    use lightning::ln::peer_handler::{IgnoringMessageHandler, MessageHandler, SimpleArcPeerManager};
+    use lightning::routing::network_graph::{NetGraphMsgHandler, NetworkGraph};
+    use lightning::util::config::UserConfig;
+    use lightning::util::events::{Event, EventHandler};
+    use lightning::util::ser::ReadableArgs;
+    use lightning_background_processor::BackgroundProcessor;
+    use lightning_net_tokio::SocketDescriptor;
+    use lightning_persister::FilesystemPersister;
+    use rand::RngCore;
+    use rpc::v1::types::H256;
+    use script::{Builder, SignatureVersion};
+    use std::cmp::Ordering;
+    use std::convert::{TryFrom, TryInto};
+    use std::net::{IpAddr, Ipv4Addr};
+    use std::sync::Arc;
+    use std::time::SystemTime;
+    use tokio::net::TcpListener;
+}
+
+cfg_native! {
+    const CHECK_FOR_NEW_BEST_BLOCK_INTERVAL: u64 = 60;
+    const BROADCAST_NODE_ANNOUNCEMENT_INTERVAL: u64 = 60;
+    const TRY_RECONNECTING_TO_NODE_INTERVAL: u64 = 60;
+}
+
+lazy_static! {
+    pub static ref NO_SUCH_TRANSACTION_ERROR: Json = json!({
+        "code": 2,
```

Comment:
Despite Lightning is not available for KMD, we should still consider that the message might be changed. For example, I got this error from komodod: `{"jsonrpc": "2.0", "error": {"code": 2, "message": "daemon error: DaemonError({'code': -5, 'message': 'No information available about transaction'})"}, "id": "0"}`. It seems that it's better to rely on error code only. Does `2` uniquely identify the "missing transaction" error?

Reply:
Error code `2` is for any electrum DaemonError. `'code': -5` is the error code that's related to a missing transaction and most daemons use the same code from what I have seen (like in the 2 cases we have here where `KMD` and `BTC` use -5), but the exact case for `transaction_unconfirmed` for BTC is the `'No such mempool or blockchain transaction'` as shown here https://github.com/bitcoin/bitcoin/blob/dca9ab48b80ff3a7dbe0ae26964a58e70d570618/src/rpc/rawtransaction.cpp#L241-L243

I can of course check for if the `"message"` contains `'code': -5` which is what I did at the start before refining, this should cause no problem as we don't provide the parameters for `getrawtransaction` such as `blockindex` or `txindex` that will cause the daemon to return a different `'code': -5` error.